Modify task_pool to allow for repeated tasks that run alongside regular tasks.

Smackteo opened the pull request:
#### cafffd84ce641f66540dacabbfd898cc19e7e3ca
<pre>
Modify task_pool to allow for repeated tasks that run alongside regular tasks.
<a href="https://rdar.apple.com/164584838">rdar://164584838</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=302420">https://bugs.webkit.org/show_bug.cgi?id=302420</a>

Reviewed by Jonathan Bedard.

repeated tasks must be added to the taskpool instead of requeued,if they were requeued they would not repeat on a specific worker.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/task_pool.py:
(_Message.__call__):
(_Task.__init__):
(_Result.__init__):
(_Result.__call__):
(_StopRepeat):
(_StopRepeat.__init__):
(_StopRepeat.__call__):
(_Process):
(_Process.main):
(TaskPool):
(TaskPool.do):
(TaskPool.wait):
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/task_pool_unittest.py:
(simple_repeat_task):
(printing_repeat_task):
(TaskPoolUnittest.test_mutually_exclusive_group_hang):
(TaskPoolUnittest):
(TaskPoolUnittest.test_repeat_task_with_groups):
(TaskPoolUnittest.test_repeat_task_multiple_groups):
(TaskPoolUnittest.test_repeat_task_with_regular_grouped_tasks):
(TaskPoolUnittest.test_repeat_task_pool_state_tracking_forked):

Canonical link: <a href="https://commits.webkit.org/304186@main">https://commits.webkit.org/304186@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c585de3c9a3d3dd6adfc7b5942e27e4dd23d1806

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134861 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/7289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46106 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142373 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4977bc0e-56bd-4a5c-bff2-6da8828edddd) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136730 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7904 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/7136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/cbbc325d-da58-41f3-a3ee-7c7029ddec4a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137807 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/5580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/120877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83887 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d1138937-6168-470e-9f3c-555df7c3c3e2) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/134209 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/5409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2966 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/39028 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145071 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6964 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/39604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/134288 "Passed tests") | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/7023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/7136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111770 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28347 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/117156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/60894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/7011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/35332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/6795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/7031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/6905 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

gsnedders reviewed (commented):
I don't understand what's meant by "infinite" here. Is the notable semantic here just that the task gets repeated until it wants to stop (which also implies it is potentially finite)? If that's the semantic here, then maybe we should just call it "repeat".

If that's all that's happening here, do we really gain that much from making such large changes, versus making `TaskPool.do` add a second callback which then re-enqueues the task? I guess this is made somewhat difficult by the fact that we only support a single callback per task… I still really want to re-write TaskPool to use `concurrent.futures` rather than implementing all the logic ourselves.

JonWBedard commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/task_pool_unittest.py:
```diff
@@ -297,3 +314,132 @@ def test_mutually_exclusive_group_hang(self):
 
             self.assertEqual(len(captured.stdout.getvalue().splitlines()), 26)
+
+        def test_repeat_task_basic(self):
```
Don't we already have a test for this? These tasks aren't actually repeated.

JonWBedard commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/task_pool_unittest.py:
```diff
@@ -297,3 +314,132 @@ def test_mutually_exclusive_group_hang(self):
+
+        def test_repeat_task_pool_state_tracking_forked(self):
+            """Test that TaskPool correctly tracks repeat task state in forked mode."""
+            with TaskPool(workers=1, force_fork=True) as pool:
```
How are repeat tasks supposed to work when `workers=1`? It doesn't seem like they could.

Smackteo replied:
oops! Fixing

JonWBedard commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/task_pool_unittest.py:
```diff
@@ -297,3 +314,132 @@ def test_mutually_exclusive_group_hang(self):
+
+                pool.wait()
+
+    def test_repeat_task_no_fork(self):
```
None of the tests have `workers=2`, which is the whole point of the `repeat=True`

JonWBedard commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/task_pool_unittest.py:
```diff
@@ -297,3 +309,161 @@ def test_mutually_exclusive_group_hang(self):
+                pool.do(action, 'regular')
+
+                # Add repeat task - in no-fork mode, this should be handled specially
+                # Since there's no actual repeat loop in no-fork mode, it should execute once
```
Is this the right behavior? Or should this be an exception?

JonWBedard commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/task_pool_unittest.py:
```diff
@@ -297,3 +309,161 @@ def test_mutually_exclusive_group_hang(self):
+                    self.assertEqual(pool.repeat_pending_count, 0)
+                    self.assertFalse(pool._has_repeat_tasks)
+
+                    pool.do(simple_repeat_task, repeat=True)
```
Seems like this should be an exception rather than special behavior.

JonWBedard commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/task_pool_unittest.py:
```diff
@@ -297,3 +309,161 @@ def test_mutually_exclusive_group_hang(self):
+            repeat_count = stdout_output.count('REPEAT_MARKER_REGULAR')
+            self.assertGreater(repeat_count, 1, "Repeat task should execute multiple times (got {})".format(repeat_count))
+
+        def test_repeat_task_internal_attributes(self):
```
This test doesn't seem very valuable, we should remove it.

Smackteo committed:
Modify task_pool to allow for repeated tasks that run alongside regular tasks.
rdar://164584838
https://bugs.webkit.org/show_bug.cgi?id=302420

Reviewed by Jonathan Bedard.

repeated tasks must be added to the taskpool instead of requeued,if they were requeued they would not repeat on a specific worker.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/task_pool.py:
(_Message.__call__):
(_Task.__init__):
(_Result.__init__):
(_Result.__call__):
(_StopRepeat):
(_StopRepeat.__init__):
(_StopRepeat.__call__):
(_Process):
(_Process.main):
(TaskPool):
(TaskPool.do):
(TaskPool.wait):
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/task_pool_unittest.py:
(simple_repeat_task):
(printing_repeat_task):
(TaskPoolUnittest.test_mutually_exclusive_group_hang):
(TaskPoolUnittest):
(TaskPoolUnittest.test_repeat_task_with_groups):
(TaskPoolUnittest.test_repeat_task_multiple_groups):
(TaskPoolUnittest.test_repeat_task_with_regular_grouped_tasks):
(TaskPoolUnittest.test_repeat_task_pool_state_tracking_forked):

Canonical link: https://commits.webkit.org/304186@main
